userspace/libsinsp/threadinfo.cpp

```diff
@@ -767,6 +768,49 @@ uint64_t sinsp_threadinfo::get_fd_limit()
 	return get_main_thread()->m_fdlimit;
 }
 
+void sinsp_threadinfo::traverse_parent_state(std::function<bool (sinsp_threadinfo *)> visitor)
+{
+	// Use two pointers starting at this, traversing the parent
+	// state, at different rates. If they ever equal each other
+	// before slow is NULL there's a loop.
+
+	sinsp_threadinfo *slow=this, *fast=this;
+
+	while(slow)
+	{
+		// Advance slow one step and call the visitor
+		slow = slow->get_parent_thread();
```

Comment:
I don't think there should be any expectation that if there's a loop, all the nodes in the loop are visited at least or exactly once. I think the goal should be to visit as much as you can while possibly detecting a loop and exit as soon as you detect a loop.